### ISSUE TITLE
Artifact Reference Subpath

### ISSUE THREAD

**Issue by AntoineDao** [argoproj/argo#3061]
# Summary

We should allow users to specify a `subpath` witthin a referenced artifact.

# Motivation

We run workflows where we generate a folder of artifacts we then want to loop over. Our current implementation is to create a hard coded S3 key where we store the output artifacts in an unzipped `folder` and then run a loop that uses the `{{item}}` to specify a subkey in the S3 artifact path.

This becomes complex when we nest these loops inside of multiple DAGs and as such it would be helpful to be able to add a subpath to a referenced artifact.

# Proposal

How do you think this should be implemented?

I think an example API could use a `fromSubpath` key to indicate when a user wants to use a subpath from a referenced artifact. Here is a rework of the classic `artifact-passing` workflow:

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: artifact-passing-
spec:
  entrypoint: artifact-example
  templates:
  - name: artifact-example
    steps:
    - - name: generate-artifact
        template: whalesay
    - - name: consume-artifact
        template: print-message
        arguments:
          artifacts:
          - name: message
            from: "{{steps.generate-artifact.outputs.artifacts.hello-art}}"
            fromSubpath: 'hello_world.txt'

  - name: whalesay
    container:
      image: docker/whalesay:latest
      command: [sh, -c]
      args: ["sleep 1; cowsay hello world | tee /tmp/hello_world.txt"]
    outputs:
      artifacts:
      - name: hello-art
        path: /tmp
        archive:
          none: {}

  - name: print-message
    inputs:
      artifacts:
      - name: message
        path: /tmp/message
    container:
      image: alpine:latest
      command: [sh, -c]
      args: ["cat /tmp/message"]
```
---

**Message from the maintainers**:

If you wish to see this enhancement implemented please add a 👍 reaction to this issue! We often sort issues this way to know what to prioritize.

**Comment by alexec** [comment]
I'm trying to understand this. You have a step that generates a directory of files, and for each file in that directory, you want to loop over it? Could you please attach your current YAML?

**Comment by sarabala1979** [comment]
I would recommend resolving the `from` variable with path concatenate in runtime. Controller is already resolving variables in other places like 'args' and command. 
```
args: ["echo result was: {{inputs.parameters.message}}"]
```

eg:
```
    - - name: consume-artifact
        template: print-message
        arguments:
          artifacts:
          - name: message
            from: "{{steps.generate-artifact.outputs.artifacts.hello-art}}/hello_world.txt"
```

**Comment by Ark-kun** [comment]
Can you use normal output artifacts (with auto-generated keys) and just make an output for every file you want to output?

```yaml
  - name: whalesay
    container:
      image: docker/whalesay:latest
      command: [sh, -c]
      args: ["cowsay hello world > /tmp/hello_world.txt && cowsay bye world > /tmp/bye_world.txt"]
    outputs:
      artifacts:
      - name: hello-art
        path: /tmp/hello_world.txt
      - name: bye-art
        path: /tmp/bye_world.txt
```

**Comment by AntoineDao** [comment]
A bit of context might help here. I am using Argo as a workflow engine, however I define the workflows and workflow templates using a separate (custom) tool. This tool lets me define re-usable/composable `container templates` and `dag templates` that I can inject into an Argo Workflow.

One of the reasons I use Argo is because it allows me to run tasks in parallel over many machines. To run these tasks in parallel I often split a big file into smaller files and iterate over them. In some cases the process of splitting files requires a DAG to perform "complex splitting".

When I nest a DAG inside another one I make use of the DAG template `outputs` key to expose any artifacts/parameters to the tasks in the "parent" DAG.  Once I do this it becomes difficult to resolve the path to an artifact output by a DAG task that is itself a DAG because the path to the artifact is defined deep inside the child DAG.

It's messy to explain so maybe this example will help:

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: loops-nested-dag-artifact-passing-
spec:
  entrypoint: parent-dag
  templates:
  - name: parent-dag
    inputs:
      parameters:
      - name: split-by
        value: 4
      artifacts:
      - name: myfile
        raw:
          data: |
            1
            2
            3
            4
    dag:
      tasks:
      - name: split-dag
        template: child-dag
        arguments:
          parameters:
          - name: split-by
            value: '{{inputs.parameters.split-by}}'
          artifacts:
          - name: file
            from: '{{inputs.artifacts.myfile}}'
      
      - name: loop-read
        dependencies: [split-dag]
        template: read-file
        withParam: '{{tasks.split-dag.outputs.parameters.file-list}}'
        arguments:
          artifacts:
          - name: file
            from: '{{tasks.split-dag.outputs.artifacts.folder}}'
            fromSubPath: '{{item}}'
     

  - name: child-dag
    inputs:
      parameters:
      - name: split-by
      artifacts:
      - name: file
    dag:
      tasks:
      - name: split
        template: split-file
        arguments:
          parameters:
          - name: split-by
            value: '{{inputs.parameters.split-by}}'
          - name: output-artifact-folder
            value: '{{workflow.name}}/split-file'
          artifacts:
          - name: file
            from: '{{inputs.artifacts.file}}'
      
      - name: list-dir
        template: directory-list
        dependencies: [split]
        arguments:
          artifacts:
          - name: directory
            from: '{{tasks.split.outputs.artifacts.split-folder}}'
    outputs:
      parameters:
      - name: file-list
        valueFrom:
          parameter: '{{tasks.list-dir.outputs.parameters.directory-file-list}}'
      artifacts:
      - name: folder
        from: '{{tasks.split.outputs.artifacts.split-folder}}'

  - name: split-file
    inputs:
      parameters:
      - name: split-by
      - name: output-artifact-folder
      artifacts:
      - name: file
        path: /cowsay/file
    container:
      image: docker/whalesay:latest
      command: [sh, -c]
      args: ["mkdir tmp && split -n {{inputs.parameters.split-by}} file tmp/ && ls tmp | tee file-list.txt"]
    outputs:
      artifacts:
      - name: split-folder
        path: /cowsay/tmp
        archive:
          none: {}
        s3:
          endpoint: minio:9000
          bucket: my-bucket
          insecure: true
          accessKeySecret:
            name: my-minio-cred
            key: accesskey
          secretKeySecret:
            name: my-minio-cred
            key: secretkey
          key: '{{inputs.parameters.output-artifact-folder}}'

  - name: directory-list
    inputs:
      artifacts:
      - name: directory
        path: /tmp/artifact
    script:
      image: python:alpine3.6
      command: [python]
      source: |
        import json
        import os
        ls = os.listdir('/tmp/artifact')
        with open('tmp/list.json', 'w') as f:
            json.dump(ls, f)
    outputs:
      parameters:
      - name: directory-file-list
        valueFrom:
          path: '/tmp/list.json'

  - name: read-file
    inputs:
      artifacts:
      - name: file
        path: /cowsay/file.txt
    container:
      image: docker/whalesay:latest
      command: [sh, -c]
      args: ["cat file.txt"]
```

The `child-dag` template creates a folder with files inside. It is impractical to point to each file individually as an artifact as the number of files depends on the `split-by` value. Similarly it is challenging to hard-code the location of the folder from outside the `child-dag` context because it is defined as a variable called `output-artifact-folder` inside the `child-dag`.

In this case it becomes useful to let Argo do what it does best (resolve the artifact path at run time) and give it a subpath to iterate over the precise file I am interested in.

**Comment by alexec** [comment]
@sarabala1979 do you want to assist as you know this well?

**Comment by alexec** [comment]
Available for testing in [v2.10.0-rc1](https://github.com/argoproj/argo/releases/tag/v2.11.0-rc1).
